mzLib/MassSpectrometry/Deconvolution/Algorithms/SpectralDeconvolutionAlgorithm.cs

```diff
@@ -0,0 +1,281 @@
+
+            FindLibraryEnvelopes();
+            IndexEnvelopes();
+            Scorer = new Scorer(Scorer.ScoringMethods.SpectralContrastAngle, PpmTolerance);
+
+        }
+
+        public override IEnumerable<IsotopicEnvelope> Deconvolute(MzSpectrum spectrum)
+        {
+
+            if (spectrum == null || spectrum.Size == 0)
+            {
+                yield break;
+            }
+
+            // For each charge state (key) store the indices corresponding to every potential isotopic envelope (value)
+            Dictionary<int, List<MinimalSpectrum>> potentialEnvelopes = FindPotentialEnvelopes(spectrum);
+
+            // iterate through charge states (potentially not necessary/performant. Could flatten)
+            foreach (var keyValuePair in potentialEnvelopes)
+            {
+                int chargeBinIndex = keyValuePair.Key - SpectralParams.MinAssumedChargeState;
+                // iterate through potential envelopes
+                foreach (var experimentalSpectrum in keyValuePair.Value)
+                {
+                    double mostAbundantMz =  experimentalSpectrum.MostAbundantMz;
+                    int massBinIndex = (int)Math.Floor((mostAbundantMz - SpectralParams.ScanRange.Minimum) *
+                                                   SpectralParams.BinsPerDalton);
+                    if (!IndexedLibrarySpectra[massBinIndex, chargeBinIndex].IsNotNullOrEmpty()) continue; // continue if there are no corresponding library spectra
+
+                    int? bestMatchListPosition = null;
+                    int currentListPosition = 0;
+                    double bestFoundScore = Scorer.PoorScore;
+                    // Score against matching theoretical envelopes
+                    foreach (MinimalSpectrum theoreticalSpectrum in IndexedLibrarySpectra[massBinIndex, chargeBinIndex])
+                    {
+                        // TODO: Rename to FindBestScore
+                        if (Scorer.TestForScoreImprovement(
+                                Scorer.Score(experimentalSpectrum,theoreticalSpectrum),
+                                bestFoundScore,
+                                out double betterScore)
+                            )
+                        {
+                            bestMatchListPosition = currentListPosition;
+                            bestFoundScore = betterScore;
+                        }
+                        currentListPosition++;
+                    }
+
+                    if (bestMatchListPosition.HasValue && 
+                        SpectrumIndexToPwsmMap.TryGetValue((massBinIndex, chargeBinIndex, (int)bestMatchListPosition), out var pwsmMatch))
+                    {
+                        yield return new IsotopicEnvelope(experimentalSpectrum, pwsmMatch, bestFoundScore);
+                    }
+                    else
+                    {
+                        //TODO: Add some averagine bullshit here
```

Comment:
Haha this one's almost worth framing

Reply:
Do you still need to add something here?

Reply:
Not on this PR. That would come in subsequent iterations